File: cpp/lib/p_1079.cpp
#include "p_1079.h"

int P_1079::Solution::numTilePossibilities(string tiles) {
  array<int, 26> freq{};
  for(const auto& c: tiles) {
    freq[c - 'A'] += 1;
  }
  int result = 0;
  backtrack(freq, result);

  return result;
}

void P_1079::Solution::backtrack(span<int> freq, int& result) {
  for(int i = 0; i < 26; i++) {
    if (freq[i] > 0) {
      result += 1;
      freq[i] -= 1;

      backtrack(freq, result);

      freq[i] += 1;
    }
  }
}

File: javascript/src/problems/p_1079.ts
export function numTilePossibilities(tiles: string): number {
  const freq = new Array(26).fill(0);

  for (const c of tiles) {
    freq[c.charCodeAt(0) - "A".charCodeAt(0)]++;
  }

  let result = 0;

  const backtrack = () => {
    for (let i = 0; i < freq.length; i++) {
      if (freq[i] > 0) {
        result++;
        freq[i]--;
        backtrack();
        freq[i]++;
      }
    }
  };

  backtrack();

  return result;
}
